### ISSUE TITLE
Support for structured access log

### ISSUE THREAD

**Issue by AndreLouisCaron** [aio-libs/aiohttp#855]
I'm trying to get my new aiohttp webapp to use structured logging based on the amazing [structlog](https://structlog.readthedocs.io/).  My app uses this all over the code base and I'd like to send the aiohttp server's access logs to the same stream.

Currently, aiohttp provides a way to customize the log format, but it always formats this as a string.  While it does format it in a popular format that a lot of tools can parse, I kind of find it annoying to have to render the fields into a string and then decode the string to extract the fields we already had in memory.

Any chance we can supply an access log object that accepts the original data?

**Comment by asvetlov** [comment]
Sorry, I don't follow.
What exact change are you requesting?

**Comment by ludovic-gasc** [comment]
We use also structured logging, but with journald and do queries with journalctl.

Instead of to capture a log line, the idea is to capture a structured log (= a dict), where each key can be used to retrieve easily information by a computer, because logs aren't always consistent for a simple grep.
And in fact, when you generate a log line, you already have a data structure in your source code because each value you show in log come from different variables.
The idea is to keep this metadata to help to find information in logs.

The less intrusive way I've found in existing logging code is to pass logging parameters by name instead of position.

Example, instead of to write:

``` python
LOG = logging.getLogger(__name__)

LOG.debug('Receive a read_all request from "%s" account on "%s"', (account_id, request.path))
```

We write: 

``` python
LOG.debug('Receive a read_all request from "%(account_id)s" account on "%(http_path)s"',
              {'request_id': request.request_id,
               'account_id': account_id,
               'aiohttp_request': request,
               'request_start': True})
```

And after, I've custom Handlers to read into aiohttp_request:

``` python
import collections
from logging import StreamHandler as _OfficialStreamHandler
from logging.handlers import SysLogHandler as _OfficialSysLogHandler
from logging.handlers import SMTPHandler as _OfficialSMTPHandler
from systemdream.journal.handler import JournalHandler as _OfficialJournalHandler

def log_record_enrichment(record):
    if record.args and isinstance(record.args, collections.Mapping):
        if 'aiohttp_request' in record.args:
            record.args['http_method'] = record.args['aiohttp_request'].method
            record.args['http_path'] = record.args['aiohttp_request'].path
            record.args['http_headers'] = str(record.args['aiohttp_request'].headers)
            del(record.args['aiohttp_request'])
        if 'aiopg_cursor' in record.args:
            record.args['pg_query'] = record.args['aiopg_cursor'].query.decode('utf-8')
            record.args['pg_status_message'] = record.args['aiopg_cursor'].statusmessage
            record.args['pg_rows_count'] = record.args['aiopg_cursor'].rowcount
            del(record.args['aiopg_cursor'])

class JournalHandler(_OfficialJournalHandler):
    # Tip: on a system without journald, use socat to test:
    # socat -v UNIX-RECV:/run/systemd/journal/socket STDIN
    def emit(self, record):
        log_record_enrichment(record)
        super().emit(record)

class StreamHandler(_OfficialStreamHandler):
    def emit(self, record):
        log_record_enrichment(record)
        super().emit(record)

class SysLogHandler(_OfficialSysLogHandler):
    def emit(self, record):
        log_record_enrichment(record)
        super().emit(record)

class SMTPHandler(_OfficialSMTPHandler):
    def emit(self, record):
        log_record_enrichment(record)
        super().emit(record)
```

As you can see, I can use standard Handlers, it isn't mandatory to use only journald.

And after, we can query logs via journalctl.
For example, if I want to see all POST for a specific endpoint:

```
journalctl HTTP_PATH=/v3.0/users HTTP_METHOD=PUT
```

It seems to be overkill compare to several grep commands, but journalctl is more like a SQL query tool for logs than a grep remplacement.
You can also generate easily statistics usage of your source code, because journald stores also the file name and the line that generates the log.

Concretely, in aiohttp, if it could be possible to pass logs values with dict instead of tuple, it means that everybody can catch the dict with a specific handler to generate directly structured logs, and no disturb for people that don't use that.

**Comment by AndreLouisCaron** [comment]
> What exact change are you request?

I was hoping to get a way to obtain the access log data to some kind of callback in order to log it myself (not as a string).

I ended up writing my own middleware to do it.  Something like this:

``` python
async def access_log_middleware(app, handler):
    """Log each request in structured event log."""

    event_log = app.get('event_log') or structlog.get_logger()
    clock = app.get('clock') or timeit.default_timer

    # Keep the request arrival time to ensure we get intuitive logging of
    # events.
    arrival_time = datetime.utcnow().replace(tzinfo=timezone.utc)

    async def access_log(request):
        ref = clock()
        try:
            response = await handler(request)
            event_log.info(
                'http.access',
                path=request.path,
                outcome=response.status,
                duration=(clock()-ref),
                request=request.get('x-request-id', '?'),
                **{'@timestamp': arrival_time},
            )
            return response
        except aiohttp.web.HTTPException as error:
            event_log.info(
                'http.access',
                path=request.path,
                outcome=error.status,
                duration=(clock()-ref),
                request=request.get('x-request-id', '?'),
                **{'@timestamp': arrival_time},
            )
            raise
        except Exception:
            event_log.info(
                'http.access',
                path=request.path,
                outcome=500,
                duration=(clock()-ref),
                request=request.get('x-request-id', '?'),
                **{'@timestamp': arrival_time},
            )
            raise

    return access_log
```

**Comment by asvetlov** [comment]
Fixed by #1303 -- `extra` is passed now
